cmd/agent/app/reporter/grpc/builder.go

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2018-2019 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpc
+
+import (
+	"errors"
+	"strings"
+
+	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/balancer/roundrobin"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/naming"
+	"google.golang.org/grpc/resolver"
+	"google.golang.org/grpc/resolver/manual"
+)
+
+// ConnBuilder Struct to hold configurations
+type ConnBuilder struct {
+	// CollectorHostPorts is list of host:port Jaeger Collectors.
+	CollectorHostPorts []string `yaml:"collectorHostPorts"`
+
+	// Resolver is custom resolver provided to grpc client balancer
+	Resolver naming.Resolver
```

Comment:
It is difficult to judge if this change achieves our end goal of hooking up custom discovery service. For example, in tchannel/builder you will see
```go
	return peerlistmgr.New(peers, b.discoverer, b.notifier,
		peerlistmgr.Options.MinPeers(defaultInt(b.DiscoveryMinPeers, defaultMinPeers)),
		peerlistmgr.Options.Logger(logger),
		peerlistmgr.Options.ConnCheckTimeout(b.ConnCheckTimeout),
	)
```
which serves in a similar role as naming.Resolve, but using the two oss APIs `b.discoverer, b.notifier` which we already implement via UNS in the internal repo. It's fine to be accepting grpc `naming.Resolver` here, but I would like to see an implementation of that expressed via `b.discoverer, b.notifier`, with some unit tests demonstrating the load balancing (similar to peerlistmgr pkg).

Reply:
I think my question here is that for tchannel the flow is working like this:
we have unsWatcher that dispatches tchannel peers to `peerListMgr.discoCh` which tchannel can use to update its peers without constructing a new connection.

![image](https://user-images.githubusercontent.com/7311035/55806520-65c92080-5aae-11e9-9e65-f293526d7bcf.png)

Based on the interface definition for `resolver` and `watcher` in grpc-go, my understanding is that as long as we are implementing a resolver and watcher based on the interface grpc library will update list of peers so it's not like tchannel-go that you need to update "peers" manually? Therefore we no longer need to implement the whole `peerListMgr` alike component?

Reply:
@black-adder Do you have any input on this while yuri is out?

Reply:
I am not saying you need peer mgr, but gRPC Resolver/Watcher are conceptually very similar to Discoverer/Notifier, so we need to implement the former in terms the latter.

Reply:
sure. I can implement a fake revolser/watcher in github for testing. I already have another story on going to provide them from internal repo for hooking up our discovery service.

Reply:
before you implement "fake" ones, check if peer mgr module already has them in the tests

Reply:
will do.